src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1261,63 +1268,86 @@ private AuthenticationResult localFlow(CallbackHandler callbackHandle,
             return cachedItem;
         }
 
+        // Trying to find refresh token first, if existed, will try to use the refresh token. 
         Logger.v(TAG, "Checking refresh tokens");
         RefreshItem refreshItem = getRefreshToken(request);
+        AuthenticationResult authResult = null;
         if (!promptUser(request.getPrompt()) && refreshItem != null
                 && !StringExtensions.IsNullOrBlank(refreshItem.mRefreshToken)) {
             Logger.v(TAG, "Refresh token is available and it will attempt to refresh token");
-            return refreshToken(callbackHandle, activity, useDialog, request, refreshItem, true);
-        } else {
-            Logger.v(TAG, "Refresh token is not available");
+            try {
+                authResult = getTokenWithRefreshToken(activity, useDialog, request, refreshItem, true);
+            } catch (AuthenticationException authenticationException) {
+                callbackHandle.onError(authenticationException);
+                return null;
+            }
+            
+            if (authResult != null && !StringExtensions.IsNullOrBlank(authResult.getAccessToken())) {
+                callbackHandle.onSuccess(authResult);
+                return authResult;
```

Comment:
We'll need more refactor in the code to make this goal, that's beyond current scope.